tox.ini

```diff
@@ -0,0 +1,82 @@
+[tox]
+minversion = 3.14.0
+envlist =
+    fix_lint
+    py38
+    py37
+    py36
+    py35
+    py27
+    pypy
+    pypy3
+    cov
+    docs
+isolated_build = true
+skip_missing_interpreters = true
+
+[testenv]
+description = run tests with {basepython}
+setenv =
+    COVERAGE_FILE = {toxworkdir}/.coverage.{envname}
+    PIP_DISABLE_PIP_VERSION_CHECK=1
+    {py27,pypy}: PYTHONWARNINGS=ignore:DEPRECATION::pip._internal.cli.base_command
+extras = test
+commands =
+    pytest \
+    --cov {envsitepackagesdir}/build --cov-context=test \
+    --cov-report=html:{envtmpdir}/html \
+    --cov-report=xml:{toxworkdir}/coverage.{envname}.xml \
+    tests {posargs}
+
+[testenv:cov]
+description = combine coverage data
+deps =
+    coverage >= 5.0.1
+    diff-cover >= 3
+extras =
+skip_install = True
+passenv = DIFF_AGAINST
+setenv =
+    COVERAGE_FILE={toxworkdir}/.coverage
+commands =
+    python -m coverage combine
+    python -m coverage report --skip-covered --show-missing
+    python -m coverage xml -o {toxworkdir}/coverage.xml
+    python -m coverage html -d {toxworkdir}/htmlcov
+    python -m diff_cover.diff_cover_tool --compare-branch {env:DIFF_AGAINST:origin/master} {toxworkdir}/coverage.xml
+depends =
+    py38
+    py37
+    py36
+    py35
+    py27
+    pypy
+    pypy3
+parallel_show_output = True
+
+[testenv:docs]
+basepython = python3.8
+description = build documentation
+extras = doc
+passenv = *
+commands =
+    sphinx-build -d "{envtmpdir}/doctree" docs/source "{toxworkdir}/docs_out" --color -b html {posargs}
+    python -c 'import pathlib; print("documentation available under file://\{0\}".format(pathlib.Path(r"{toxworkdir}") / "docs_out" / "index.html"))'
+
+[testenv:fix_lint]
+description = format the code base to adhere to our styles, and complain about what we cannot do automatically
+basepython = python3.8
+passenv = *
+deps = pre-commit >= 2
+skip_install = True
+commands =
+    pre-commit run --all-files --show-diff-on-failure
+    python -c 'import pathlib; print("hint: run \{\} install to add checks as pre-commit hook".format(pathlib.Path(r"{envdir}") / "bin" / "pre-commit"))'
```

Comment:
Which part?

Reply:
Inlining python in `python -c` to print a message. Doesn't tox have a better way to deal with this? I do not think inlining escaped python inside a tox file is a great UX. This feels very hacky.

Reply:
It does not. tox tries to not do too much magic that can be done with already available tools. And the easiest way to print something in a cross-platform fashion (assuming you have python) is to use the print statement in Python. I can see how at first this might seem odd, but you're the first person to complain about it. There's no better way, of course I can remove the print, but feels like provides some value 🤷

Reply:
I would say printing is a common need in a tool like tox, you do it here several times. Forcing people to `python -c 'print("something")'` is a very bad UX IMO. It works, so maybe that's why nobody complained.

Reply:
tox, in general, does not want to be a wrapper around already available methods and tries to avoid doing stuff it can. Printing is one of these. Noted that this is bad UX, but that does not change the fact that there's no better way available today. And to pronounce with a tox maintainer on my head now it's unlikely to change anytime soon.